File: oc/CustomCell/CustomCell/controller/ViewController.m
//
//  ViewController.m
//  CustomCell
//
//  Created by dfang on 2017-12-12.
//  Copyright © 2017年 com.east.customcell. All rights reserved.
//

#import "ViewController.h"
#import "MicroBlogCell.h"
#import "MicroBlogHeaderView.h"
#import "MicroBlogFooterView.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate, MicroBlogFooterViewDelegate>

@property (nonatomic, strong) NSArray *microBlogList;
@property (nonatomic, strong) UITableView *tableView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self.view addSubview:self.tableView];
    
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    
    self.tableView.tableHeaderView = [[MicroBlogHeaderView alloc] initWithHeaderHeight:140];
    MicroBlogFooterView *footerView = [[MicroBlogFooterView alloc] initWithFooterHeight:34];
    footerView.delegate = self;
    self.tableView.tableFooterView = footerView;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSArray *)microBlogList {
    if (!_microBlogList) {
        _microBlogList = [MicroBlog microBlogList];
    }
    return _microBlogList;
}

- (UITableView *)tableView {
    if (!_tableView) {
        _tableView = [[UITableView alloc] initWithFrame:self.view.frame style:UITableViewStylePlain];
    }
    return _tableView;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    MicroBlogCell *cell = [MicroBlogCell microBlogCellWithTableView:tableView andCellStyle:UITableViewCellStyleDefault];
    cell.microBlog = self.microBlogList[indexPath.row];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.microBlogList.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    MicroBlog *microBlog = self.microBlogList[indexPath.row];
    return microBlog.rowHeight;
}

#pragma mark - method for MicroBlogFooterViewDelegate
- (void)footerViewDidClickedLoadMoreBtn:(MicroBlogFooterView *)microBlogFooterView {
    NSLog(@"call footerViewDidClickedLoadMoreBtn");
}

@end

File: swift/CustomCell/CustomCell/controller/ViewController.swift
//
//  ViewController.swift
//  CustomCell
//
//  Created by dfang on 2017-12-13.
//  Copyright © 2017年 com.east.customcell. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, MicroBlogFooterViewDelegate {
    
    lazy var tableView: UITableView = UITableView.init()
    lazy var microBlogFrameList: [MicroBlogFrame] = {
        let microBlogList = MicroBlog.microBlogList()
        var _microBlogFrameList = NSMutableArray.init(capacity: microBlogList.count)
       
        for microBlog in microBlogList {
            var microBlogFrame:MicroBlogFrame = MicroBlogFrame.init()
            microBlogFrame.microBlog = microBlog as! MicroBlog
            _microBlogFrameList.add(microBlogFrame)
        }
        
        return _microBlogFrameList as! [MicroBlogFrame]
    }()
    
    lazy var headerView:MicroBlogHeaderView = MicroBlogHeaderView.init(_height: 140)
    lazy var footerView: MicroBlogFooterView = MicroBlogFooterView.init(_height: 34)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        tableView.frame = self.view.frame
        tableView.delegate = self
        tableView.dataSource = self
        
        tableView.tableHeaderView = headerView
        tableView.tableFooterView = footerView
        footerView.delegate = self
        
        self.view.addSubview(tableView)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return microBlogFrameList.count
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let microBlogFrame:MicroBlogFrame = microBlogFrameList[indexPath.row]
        return microBlogFrame.rowHeight
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell:MicroBlogCell = MicroBlogCell.microBlogCellWithTableView(tableView: tableView)
        cell.microBlogFrame = microBlogFrameList[indexPath.row]
        return cell
    }
    
    func footerViewDidClickedLoadMoreBtn(footerView: MicroBlogFooterView) {
        print("footerViewDidClickedLoadMoreBtn clicked")
    }
}

